src/java.base/share/classes/jdk/crac/impl/AbstractContextImpl.java

```diff
@@ -22,86 +22,85 @@
 
 package jdk.crac.impl;
 
-import jdk.crac.CheckpointException;
-import jdk.crac.Context;
-import jdk.crac.Resource;
-import jdk.crac.RestoreException;
-import sun.security.action.GetBooleanAction;
+import jdk.crac.*;
+import jdk.internal.crac.LoggerContainer;
 
-import java.util.*;
-import java.util.stream.Collectors;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
 
-public abstract class AbstractContextImpl<R extends Resource, P> extends Context<R> {
+public abstract class AbstractContextImpl<R extends Resource> extends Context<R> {
+    protected List<Resource> restoreQ = null;
 
-    private static class FlagsHolder {
-        public static final boolean DEBUG =
-            GetBooleanAction.privilegedGetProperty("jdk.crac.debug");
+    protected static <E extends Exception> void recordExceptions(E source) {
+        Throwable[] suppressed = source.getSuppressed();
+        if (suppressed.length == 0) {
+            Core.recordException(source);
+        }
+        for (Throwable t : suppressed) {
+            Core.recordException(t);
+        }
     }
 
-    private WeakHashMap<R, P> checkpointQ = new WeakHashMap<>();
-    private List<R> restoreQ = null;
-    private Comparator<Map.Entry<R, P>> comparator;
+    protected void setModified(R resource, String msg) {
+        Core.recordException(new CheckpointException(
+                "Adding resource " + resource + " to " + this + (msg != null ? msg : "")));
+    }
 
-    protected AbstractContextImpl(Comparator<Map.Entry<R, P>> comparator) {
-        this.comparator = comparator;
+    protected void invokeBeforeCheckpoint(Resource resource) {
+        LoggerContainer.debug("beforeCheckpoint {0}", resource);
+        // Resource.afterRestore is invoked even if Resource.beforeCheckpoint fails
+        restoreQ.add(resource);
+        try {
+            resource.beforeCheckpoint(semanticContext());
```

Comment:
Theoretically, the method could do that. However, here the purpose of `semanticContext()` is to pass the context to which it was registered rather than the subcontext where it is stored (but this is an implementation detail that the resource does not know about).

Reply:
Just realized that this is required for PriorityContext implementation. But that is the implementation of that class, it's wrong ACI has to care about that.

Reply:
Yes, it's a bit of enforced flexibility of the base class (through allowing to override a method), though it doesn't need about the use case.

I could use just a list rather than sub-contexts but that would require duplicated code.